### PR TITLE
[WIP] use `now.sh` to host isomorphic web app example

### LINKED ISSUE #18: Put demo on heroku or equiv.
I'm willing to bet that most people don't have time to clone/run and would still love to play w/ what you've done.

https://github.com/emk/heroku-buildpack-rust

### PR DESCRIPTION
**Pull request by chinedufn**
[WORK IN PROGRESS]

Made some progress but the build kept failing and now I've hit the rate limit so will have to try again later

image

Closes #18

### CODE REVIEW

**Comment by chinedufn** [comment]
\cc @OutThisLife @deamme @qm3ster

**Comment by deamme** [comment]
My logs: https://gist.github.com/deamme/bafa164f11c5f95ca02036cab12ae2e4

**Comment by qm3ster** [comment]
Why build it at now.sh?
Why not build it locally with a docker multi-stage build then upload the minimal runtime container?

**Comment by chinedufn** [comment]
@deamme thanks! It looks like the issue is the `mkdirp`. I'm guessing you can't run bash commands after `FROM scratch` or something. I'll just try to `COPY` the three nested directories instead of using `mkdirp`.

@qm3ster sorry I'm very much a docker noob, what would that look like? Perhaps you have a link to point us in the right direction?

Thanks both!

**Comment by qm3ster** [comment]
I think the [example in docs](https://docs.docker.com/develop/develop-images/multistage-build/#use-multi-stage-builds) is pretty good:

```dockerfile
FROM golang:1.7.3
WORKDIR /go/src/github.com/alexellis/href-counter/
RUN go get -d -v golang.org/x/net/html  
COPY app.go .
RUN CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o app .

FROM alpine:latest  
RUN apk --no-cache add ca-certificates
WORKDIR /root/
COPY --from=0 /go/src/github.com/alexellis/href-counter/app .
CMD ["./app"]  
```
The idea is that you use a build container, so that the executable is built in a consistent way regardless of the host platform, then create an image which just copies the built executable into a minimal os with no build tools.
In this case, we would be able to avoid all of `rustup` and `nodejs`, and also `wasm-opt` if we begin to include it.

**Comment by chinedufn** [comment]
Hmm - isn't that what we're doing here in this PR with `FROM scratch` ?

**Comment by qm3ster** [comment]
Yes, I missed that completely.
And this isn't helping us either, since `now` is the one that builds the image.

**Comment by chinedufn** [comment]
Ah gotcha

Perhaps I'm missing something, but I don't yet see the issue with `now` building the image.

I think the main goal here is to have an easy way to show a real example of a Percy web application (our basic isomorphic web app surely needs work but it's a start ha)

So whatever allows us to do that works IMO.

In this case, if it's easy for `now` to handle the building then sure. If it's easy to build it locally and upload it then also sure. I just don't know how to do that so going with "whatever works" at the moment (;

My logging works again so I'm currently mid `now` build with my fingers crossed.

Happy to change course if you have a better approach in mind here!

**Comment by qm3ster** [comment]
Another solution would be to install all dependencies , including webpack, and then push that out as an image to Docker Hub. This will speed up the build and reduce logging output.
So the build will be

```dockerfile
FROM chinedufn/percy:latest as build
WORKDIR /usr/src
COPY . .

RUN cargo +nightly build -p isomorphic-client --release --target wasm32-unknown-unknown
RUN wasm-bindgen --no-typescript target/wasm32-unknown-unknown/release/isomorphic_client.wasm --out-dir ./examples/isomorphic/client
 # Build WASM module
# TODO: --mode=production . Need to make sure it works locally. If it doesn't try disabling mangling
RUN ./node_modules/webpack-cli/bin/cli.js --mode=development ./examples/isomorphic/client/client-entry-point.js -o ./examples/isomorphic/client/bundle.js
 # This gets around the 100Mb limit by ditching our `target` directory
FROM scratch
```
and so on

**Comment by qm3ster** [comment]
Yeah, I get that. It actually looks like `now` doesn't allow you to just deploy a built image at all.
You could of course just deploy a folder with just a dockerfile:
```dockerfile
FROM me/my-prebuilt-app:latest
```
🤷‍♂️

**Comment by chinedufn** [comment]
Ohhh nice! Ok once this is all working I'll create a new issue for creating a base image to speed up build time.

I think one thing that should come before that is automatically deploying to `now` from `CI`. Probably something we can Google for an example of. Would be great to be able to upgrade the web app example and have it reflected right away without needing to go manually run some commands.

Actually while we're on the topic.. There should probably be another issue tracking things that we want to showcase in the web app example so that we can start incrementally improving it..

LMK if you disagree!

**Comment by chinedufn** [comment]
@deamme I've fortunately fixed that issue. The latest problem is in my screenshot above. Not sure why that isn't working but currently running a build with an attempted fix -> https://travis-ci.org/chinedufn/percy

**Comment by qm3ster** [comment]
The error gives this link - https://github.com/zeit/now-cli/blob/master/errors/verification-timeout.md

Perhaps we need to do `EXPOSE` in the dockerfile to specify the port to check for liveness? Or somehow otherwise specify a readiness/health check or lack thereof?
Maybe try with `--no-verify` first?

**Comment by chinedufn** [comment]
@qm3ster hmm I think it's unable to even find the binary for some reason https://github.com/chinedufn/percy/pull/23#issuecomment-418173499

**Comment by qm3ster** [comment]
Does locally building and running the image work?

**Comment by chinedufn** [comment]
> Does locally building and running the image work?

Haven't tried will try

**Comment by chinedufn** [comment]
Ok so I built the image locally.

It's totally finding the binary and running it, the probably is that the program is generating a `no such file or directory" error.

```
$ docker run percy /isomorphic-server
standard_init_linux.go:190: exec user process caused "no such file or directory"
```

I think this means something along the realms of the binary is looking for something that isn't actually there and crashing immediately.

The [now.sh rust Docker](https://github.com/zeit/now-examples/blob/f3d2d2d716f4cd871c7e8c084dd92918b20bc606/rust-http-microservice/Dockerfile#L1) example uses [clux/muslrust](https://github.com/clux/muslrust) which from what I understand would make the binary statically link against some dependencies instead of dynamically expecting them to be there. (Someone who fully understands feel free to correct me!)

So in short.. I think we just need to get the build working `FROM clux/musl-rust` and then we're good to go.

**Comment by chinedufn** [comment]
This PR is now ready for review

---

@qm3ster @deamme automatically deploying PR branches are handled by the [now GitHub integration](https://zeit.co/github).

We use travis to deploy the `master` branch to `now.sh`

Mind reviewing the PR when you have a moment?

---

I think next steps are an issue where we toss ideas on what we want a demo to showcase and then split that up into separate issues and (hopefully) attract others to help add them!

As we add more functionality we'll run into bugs and bottlenecks and fix / document along the way.

Then eventually when the example is super solid we can take what we've learned about designing a Rust web application and build a seed project `percy new my-app-name` that people can get started with.

FWIW I'm working on a couple of private apps now using Percy - so it'd be nice to also put some stuff out in the open by enhancing the example app!

**Comment by chinedufn** [comment]
Gonna go ahead and merge this just to get the demo live and have a starting place, but lmk if you want anything changed!

**Comment by qm3ster** [comment]
1. Why are we not installing `wasm-bindgen` using `cargo`?
2. Is there a reason we are using `webpack` as opposed to, say `rollup`?
3. Why is `examples/isomorphic/client/client-entry-point.js` doing an asynchronous import of `client.js`? Do we expect to do other things while `client.js` is downloading?

**Comment by chinedufn** [comment]
1. Main reason is that building from source is very slow (a few minutes) while downloading a release binary is seconds - but hmm yeah I forgot the versions. With ether approach though we’d need to update the version to be exact.

2. No reason - just what most of the examples in the rust wasm ecosystem currently use so I started from there. What does rollup provide?

3. Also just took that from the rust wasm starter tutorials and haven’t looked at it again since.

**Comment by qm3ster** [comment]
I changed my dockerfile to this:
```dockerfile
# Used to host the isomorphic web app example

FROM rust:latest as build

# Install wasm32-unknown-unknown-target
RUN rustup default nightly
RUN rustup target add wasm32-unknown-unknown
RUN rustup target add x86_64-unknown-linux-musl

# Node.js & npm
# Build tools for making npm install work
RUN curl -sL https://deb.nodesource.com/setup_9.x | bash -
RUN apt-get install -y nodejs build-essential libssl-dev pkg-config

# Install WASM bindgen CLI
RUN cargo install wasm-bindgen-cli

# Now begins the actual build

WORKDIR /usr/src

COPY . .

# Get npm package dependencies
RUN npm install

# Build example isomorphic server binary
RUN cargo build -p isomorphic-server --release --target x86_64-unknown-linux-musl

# Compile to WASM
RUN cargo build -p isomorphic-client --release --target wasm32-unknown-unknown

# Build WASM module
# TODO: --mode=production . Need to make sure it works locally. If it doesn't try disabling UglifyJS mangling
RUN wasm-bindgen --no-typescript target/wasm32-unknown-unknown/release/isomorphic_client.wasm --out-dir ./examples/isomorphic/client
RUN ./node_modules/webpack-cli/bin/cli.js --mode=development ./examples/isomorphic/client/client-entry-point.js -o ./examples/isomorphic/client/bundle.js

# This gets around the 100Mb limit by re-starting from a tiny image
# We tried `scratch` and `alpine:rust` but targeting them proved difficult so going the easy route.
FROM scratch

# At the moment our server expects the files to be in `/examples/isomorphic/client/{filename}` so we copy the examples dir
COPY --from=build /usr/src/target/x86_64-unknown-linux-musl/release/isomorphic-server /
COPY --from=build  /usr/src/examples /examples

EXPOSE 7878/tcp

ENTRYPOINT ["/isomorphic-server"]
```
Now it caches most layers locally, enabling faster builds.

Can't actually test right now because I get
```
thread 'main' panicked at 'function not implemented on non-wasm32 targets', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/wasm-bindgen-0.2.21/src/lib.rs:415:1
```
but it seems to do do everything successfully.

This error just really confuses me since it seems bindgen-related, but there shouldn't be any bindgen in server, only client?
Happens outside docker too:
```
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:475
   5: std::panicking::continue_panic_fmt
             at libstd/panicking.rs:390
   6: rust_begin_unwind
             at libstd/panicking.rs:325
   7: core::panicking::panic_fmt
             at libcore/panicking.rs:77
   8: core::panicking::panic
             at libcore/panicking.rs:52
   9: wasm_bindgen::__wbindgen_describe_closure
             at /home/mihail/.cargo/registry/src/github.com-1ecc6299db9ec823/wasm-bindgen-0.2.21/src/lib.rs:409
  10: <wasm_bindgen::closure::Closure<T>>::wrap::breaks_if_inlined
             at /home/mihail/.cargo/registry/src/github.com-1ecc6299db9ec823/wasm-bindgen-0.2.21/src/closure.rs:159
  11: <wasm_bindgen::closure::Closure<T>>::wrap
             at /home/mihail/.cargo/registry/src/github.com-1ecc6299db9ec823/wasm-bindgen-0.2.21/src/closure.rs:163
  12: <wasm_bindgen::closure::Closure<T>>::new
             at /home/mihail/.cargo/registry/src/github.com-1ecc6299db9ec823/wasm-bindgen-0.2.21/src/closure.rs:94
  13: isomorphic_app::App::render
             at examples/isomorphic/app/src/lib.rs:50
  14: isomorphic_server::server::serve::handle_connection
             at examples/isomorphic/server/src/server.rs:60
  15: isomorphic_server::server::serve
             at examples/isomorphic/server/src/server.rs:23
  16: isomorphic_server::main
             at examples/isomorphic/server/src/main.rs:7
  17: std::rt::lang_start::{{closure}}
             at /checkout/src/libstd/rt.rs:74
  18: std::panicking::try::do_call
             at libstd/rt.rs:59
             at libstd/panicking.rs:310
  19: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:102
  20: std::rt::lang_start_internal
             at libstd/panicking.rs:289
             at libstd/panic.rs:392
             at libstd/rt.rs:58
  21: std::rt::lang_start
             at /checkout/src/libstd/rt.rs:74
  22: main
  23: __libc_start_main
  24: _start
```

**Comment by chinedufn** [comment]
Does that `cargo install wasm-bindgen-cli` get cached for subsequent `now` builds?
Asking because It's slower and it doesn't solve the version issue since once `wasm-bindgen-cli` updates the versions will be off again.

In terms of those errors, I'm running `cargo test --all` locally and it looks like tests are failing with the same error.

The error is pointing to this line:

```
// examples/isomorphic/app/src/lib.rs:50
        html! {
        <div>
          <span> { "The button has been clicked: " click_component  " times!"} </span>
          <button !onclick=move|| { state.borrow_mut().msg(Msg::Click) },>{ "Click me!" }</button>
          <div> { "In this time " click_count " rustaceans have been born." } </div>
        </div>
        }
```

It looks like `Closure::new` might no longer be implemented on non wasm targets.

Right now we store events such as `!onclick` as closures, but on the server side we just ignore them.

https://github.com/chinedufn/percy/blob/916ceaa390c9ef8a7dd9ada77bda0ffb4f5f376e/virtual-dom-rs/src/virtual_node/mod.rs#L265

So the issue here is that the newer version of wasm-bindgen no longer allows you to make closures unless you're in `wasm32` target

**Comment by qm3ster** [comment]
1. According to docs, it doesn't seem `now` currently allows any caching. Significantly faster builds on `now` can be achieved by uploading the tools section as a public image to dockerhub and only building the project, which is why I reorganized the dockerfile thusly.

2. Yeah, that makes sense. An unrelated idea about events that I was pondering was this:
What if we have a tiny js payload that buffers all input events so that they are applied once hydration happens, not ignored? The fact that currently the server only allows downloading one file at a time is a perfect showcase of how that would be beneficial on a truly slow connection: If you go to https://percy-isomorphic.now.sh/ now with no cache and start clicking, all the clicks until "Loaded..." are ignored. This way they would be applied after loading is finished.

**Comment by qm3ster** [comment]
Interesting development: Docker extension for VS Code automatically suggested me this image - https://github.com/yasuyuky/docker-rust-nightly-musl
while I was trying to switch from `rust:latest` to `rust:nightly` to save time.
It must have seen that I apt-get install musl-tools for `backtrace-sys`, an `actix` dependency.

**Comment by qm3ster** [comment]
Oh god DAMN, look at that `brotli` go! From a lighthouse score of **54** to a score of [**93**](https://percy-isomorphic-xmnntqahvr.now.sh/) :ok_hand:
